### PR TITLE
Draft API

### PR DESCRIPTION
**Pull request by DBCerigo**
[DO NOT MERGE]

This is a request for review of a drafted API for `comp`. It should not be merged. 

What's being asked of you as a reviewer (each of you as a reviewer is kinda specific to a question, so fine to just comment on just the parts most pertinent to you): 
- Read the draft and answer (some of) the following:
  - Would this package fulfil the needs and requirements of Build Up's model development and validation?
  - Would this package fulfil the needs and requirements of dvp wanting to have a repeatable structure to their model validation and development for clients?
  - Would this package fulfil the needs of a user doing a Kaggle competition, wanting to have a structured model validation and development?
  - Could this package be useful to academics/researchers for model validation and development?

Any questions are aspects that you want further clarification/more detailed interface or implementation details then do just comment and I'll add.

### CODE REVIEW

**Comment by watermosaic** [comment]
@DBCerigo Thanks for the invite. I did read this over, but I haven't had anything to do with the model dev / validation process. It's a bit beyond me in an "I don't have the prerequisite knowledge to follow" way.

**Comment by t8ch** [comment]
@DBCerigo hey, unfortunately I am also lacking some background knowledge; mostly regarding your general model train/test pipeline and the design of `truman`. Would it be a good idea to depict here on a high level where `comp` fits into to the project's workflow/pipeline?

On another note, I don't see exactly where data enter the picture here (except `data_limit` argument). I assume that's because the validation data set is fixed (not "mutable") and will be used by a validation function. Though couldn't it be that the validation data changes over time, e.g. because there were erroneous samples or the dataset can be increased? If so it might be nice to also register and "version" the data. [Please ignore if I am having a major misunderstanding.]

**Comment by DBCerigo** [comment]
@t8ch thanks for taking the time to put down your thoughts, v helpful!

> @DBCerigo hey, unfortunately I am also lacking some background knowledge; mostly regarding your general model train/test pipeline and the design of `truman`. Would it be a good idea to depict here on a high level where `comp` fits into to the project's workflow/pipeline?

Yep, I'll stubs for an example `model` and `validation` into the readme of the end product I think, hopefully should make this clearer. 

> On another note, I don't see exactly where data enter the picture here (except `data_limit` argument). I assume that's because the validation data set is fixed (not "mutable") and will be used by a validation function. Though couldn't it be that the validation data changes over time, e.g. because there were erroneous samples or the dataset can be increased? If so it might be nice to also register and "version" the data. [Please ignore if I am having a major misunderstanding.]

Yea good questions. The intention is that everything that pertains to a specific `validation` is specified and controlled within it - this any training data, or anything/everything at all that is needed to _fully specify_ a `validation` that is exactly repeatable and thus can be used to compare models.  So say if the March tagging data got updated, then one would register a new `validation` that points that that updated data with an id like: 
```
topic_tagging_validation_registry.register(
    id="5-fold-data-up-to-March-v1",
    entry_point=topic.validations.KFold,
    kwargs={"folds": 5, "data_limit": "March", "data_version": "v1"}
) 
```
On the backend, i.e. data storage; this could just be pointing to different file with a storage bucket, selecting the file based on the "data_version" param. 

I guess also missing the explainers, is that this isn't meant to _enforce_ and of the rules re comparing and validating models in a robust and fair way, but is only meant to serve a structure _follow_ the rules. In the same way one can still access "private" parts of python packages that have an `__...` prefix, but it strongly indicates the caveats in doing so.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,91 @@
...
+    id="5-fold-data-up-to-June-v0",
+    entry_point=topic.validations.KFold,
+    kwargs={"folds": 5, "data_limit": "June"}
+)
+```
+
+### Run
+
+```
+comp.run(topic_tagging_validation_registry, topic_tagging_model_registry, {"output_directory": "../topic_tagging_validatin_store"}
```
Comment (ali-tny) [inline README.md:69]: 🐼 typo `validatin`

**Review by ali-tny — APPROVED** [review]
as top level summary i think it looks good - good that it's borrowing a lot from truman. i'll mull on it over next few days

**Comment by alex-hh** [comment]
I really like the idea of having something lightweight and focussing on standardising identifiers, filenames, output types etc. while being reasonably agnostic about details of model/validation procedure.

Don’t know if it is reasonable to at least allow the possibility of having quite standardised outputs? I think this is what you’re suggesting in the store section, but rapid access to objects like training time series or static final metrics in a way that is easy to compare/plot/interact with is obviously great for development and could be fairly lightweight presumably. Is this part of the idea?

Beyond that I think I could do with understanding a bit better how I’d be using the registries. Is the idea that like particular sets of models / environments would be (permanently) defined in code somewhere as basically config, and then these could be accessed in scripts? Would be quite keen to play around with this kind of set up actually. But yeah I guess I’m wondering if these are meant to be stable components that are designed to be re-used? And whether you would also envision using for sort of one-off development. e.g. say I want to do a hyper parameter selection with loads of different model configs. Do I go via a model registry? (Basically not sure how to think about what I’d want in this situation but feels different from e.g. validation of stable models).

File: README.md [hunk 2]
```diff
@@ -1,2 +1,91 @@
...
+### Store
+
+Again, will follow as implemented in `truman` except that in `comp` it will be a lot more general.
+
+A `validation` should output a table row per `model` it validates on which summarises the results for that validation-model
+combo. A `validation` may also output a more comprehensive set of results, as either a dict (stored as json) or a dataframe
+stored as a table.
+And `models` may output data on it's "history"/"internals" (haven't found good name for this), which can again be dict or dataframe,
+and would likely including detailing of the models training process and any pertinent info for understanding the model's
+functioning.
```
Comment (benjamincerigo) [inline README.md:83]: Maybe there should be an interface for the validation so that it does the storage of the details. So that this is not in comp but can be implemented by the package that is creating the validation. 

Maybe this could get the name from the general store function:
```
class Validation(Protocal):
    ...

   def store(write_base_fp: str):
       """Implemented by child."""
      ...

```

This could also be a way to do it for the models. 

The `run` or the `store` function of `truman` would then need to be changed to run these methods on the `validation` and `models`.

**Review by benjamincerigo — APPROVED** [review]
I think this all sounds great.

The interface seems clean and easy to understand. 

In my (very novice modeling) opinion this would fulfil the needs of Build Up.

Would be interested to see what you think the process for developing models using `comp` would be. 
In my mind this would be (this is defiantly wrong):
- Set up a repo
- Create a training set
- Create a validation that uses the training set
- Create a model that attempts to predict the labels in the training set
- Create a validation notebook that would create the validation and model registries and run them.
- Make CI that runs the validation notebook

**Comment by DBCerigo** [comment]
Hi everyone, big thanks for all the time taken to input your thoughts. Really really useful. 

Do checkout the other PRs I'm opening to see how the implementation is turning out :)

Will leave this PR open for a bit but will eventually close it, probably once first non-alpha release out.

**Comment by DBCerigo** [comment]
Thanks all for the input and ideas and thoughts on this! Close this PR now as v1 about ready for release.
